### PR TITLE
Wire invoice

### PR DESCRIPTION
**Pull request by benrudolph**
Ability to create _minimal_ wire invoices. Wire invoice balances cannot be adjusted; they will just be used to as a mechanism for recording a wire invoice. Billing team will have to manual reconciliation.

Some screenshots:
screen shot 2015-04-22 at 1 57 07 pm
screen shot 2015-04-22 at 1 56 57 pm
screen shot 2015-04-22 at 1 57 13 pm
screencapture-localhost-8000-hq-accounting-wire_invoices-1429725386032

This should probably go through QA before merging. One thing to note, this breaks the existing data model because it ties the `WireInvoice`s to `Domain`s which is not what the `Invoice` does (Invoice -> subscription -> subscriber -> Domain). This may be a sign that we are outgrowing the existing data model.

@nickpell @biyeun 
cc: @millerdev

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 8 errors:

* corehq/apps/accounting/invoicing.py, line 237 - E201 whitespace after '('
* corehq/apps/accounting/invoicing.py, line 240 - E124 closing bracket does not match visual indentation
* corehq/apps/accounting/tests/test_wire_invoice.py, line 5 - F401 'generator' imported but unused
* corehq/apps/accounting/tests/test_wire_invoice.py, line 7 - F401 'WireInvoice' imported but unused
* corehq/apps/domain/views.py, line 1105 - F841 local variable 'wire_invoice' is assigned to but never used
* corehq/apps/accounting/static/accounting/ko/accounting.payment_method_handler.js, line 52 - Missing semicolon.
* corehq/apps/accounting/static/accounting/ko/accounting.payment_method_handler.js, line 59 - Missing semicolon.
* corehq/apps/accounting/static/accounting/ko/accounting.payment_method_handler.js, line 200 - Missing semicolon.

File: corehq/apps/accounting/forms.py [hunk 1]
```diff
@@ -1784,29 +1785,28 @@ def __init__(self, invoice, *args, **kwargs):
             EditSubscriptionView,
             ManageBillingAccountView,
         )
+        if not invoice.is_wire:
+            subscription_link = mark_safe(make_anchor_tag(
```
Comment (nickpell) [inline corehq/apps/accounting/forms.py]: nice refactor

File: corehq/apps/accounting/interface.py [hunk 2]
```diff
@@ -406,15 +406,176 @@ def get_subtotal_and_deduction(line_items):
...
+            )
+
+        filters.update(is_hidden_to_ops=False)
+
+        return filters
+
+    @property
+    @memoized
+    def invoices(self):
+        return WireInvoice.objects.filter(**self.filters)
```
Comment (nickpell) [inline corehq/apps/accounting/interface.py]: FYI chaining filters is more common than building and passing a ```dict``` (contrary to other examples in this file) (no fix required)

File: corehq/apps/accounting/interface.py [hunk 3]
```diff
@@ -487,29 +648,22 @@ def rows(self):
...
                 format_datatables_data(
-                    mark_safe(
-                        '<a href="%(account_url)s">%(name)s</a>' % {
-                            'account_url': reverse(
-                                ManageBillingAccountView.urlname,
-                                args=[invoice.subscription.account.id]),
-                            'name': invoice.subscription.account.name,
-                        }
-                    ),
+                    mark_safe(make_anchor_tag(account_href, account_name)),
```
Comment (nickpell) [inline corehq/apps/accounting/interface.py]: :+1:

File: corehq/apps/accounting/interface.py [hunk 4]
```diff
@@ -559,19 +713,20 @@ def rows(self):
...
-                        '   data-target="#adjustBalanceModal-%(invoice_id)d"'
-                        '   href="#adjustBalanceModal-%(invoice_id)d"'
-                        '   class="btn">Adjust Balance</a>' % {
-                            'invoice_id': invoice.id
-                        }),
-                    mark_safe(
-                        '<a href="%s" class="btn">Go to Invoice</a>'
-                        % reverse(InvoiceSummaryView.urlname, args=(invoice.id,))
-                    )
+                    mark_safe(make_anchor_tag(adjust_href, adjust_name, adjust_attrs)),
```
Comment (nickpell) [inline corehq/apps/accounting/interface.py]: :+1:

File: corehq/apps/accounting/interface.py (hunk 2 above)
Comment (benrudolph) [inline corehq/apps/accounting/interface.py]: ah good to know

File: corehq/apps/accounting/models.py [hunk 5]
```diff
@@ -1359,27 +1359,92 @@ def can_reactivate_domain_subscription(cls, account, domain, plan_version,
...
 
 
-class Invoice(models.Model):
-    """
-    This is what we'll use to calculate the balance on the accounts based on the current balance
-    held by the Invoice. Balance updates will be tied to CreditAdjustmentTriggers which are tied
-    to CreditAdjustments.
-    """
-    subscription = models.ForeignKey(Subscription, on_delete=models.PROTECT)
+class InvoiceBase(models.Model):
```
Comment (nickpell) [inline corehq/apps/accounting/models.py]: This approach looks solid :)

File: corehq/apps/accounting/static/accounting/ko/accounting.payment_method_handler.js [hunk 6]
```diff
@@ -1,16 +1,72 @@
...
+var BillingHandler = function (formId, opts) {
     'use strict';
     var self = this;
-
-    self.errorMessages = errorMessages || {};
-    self.submitBtnText = submitBtnText;
-    self.form_id = form_id;
-
+    self.formId = formId;
+    self.name = name;
```
Comment (benrudolph) [inline corehq/apps/accounting/static/accounting/ko/accounting.payment_method_handler.js]: not sure where this came from. will remove

File: corehq/apps/accounting/static/accounting/ko/accounting.payment_method_handler.js [hunk 7]
```diff
@@ -80,31 +127,10 @@ var PaymentMethodHandler = function (errorMessages, submitBtnText, form_id) {
 
     self.processPayment = function () {
         if (self.costItem().isValid()) {
-            self.selectedCard().process(self.submitForm);
+           self.selectedCard().process(self.submitForm);
```
Comment (benrudolph) [inline corehq/apps/accounting/static/accounting/ko/accounting.payment_method_handler.js]: accident

**Comment by nickpell** [comment]
This looks ready for QA.

Since you're adding a billing record for wire invoices, it might make sense to add an accounting admin interface to view them, like how ```BillingRecord```s are viewable in the "Payment Record" report.

**Comment by benrudolph** [comment]
good call, will do

File: corehq/apps/accounting/forms.py [hunk 8]
```diff
@@ -1784,29 +1785,28 @@ def __init__(self, invoice, *args, **kwargs):
...
+                    end_date=subscription.date_end,
+                )
+            ))
+        else:
+            subscription_link = 'N/A'
+
         self.helper.layout = crispy.Layout(
             crispy.Fieldset(
-                'Invoice #%s' % invoice.invoice_number,
+                '{}Invoice #{}'.format('Wire' if invoice.is_wire else '', invoice.invoice_number),
```
Comment (millerdev) [inline corehq/apps/accounting/forms.py]: Should there be a space after "Wire"?

**Comment by benrudolph** [comment]
@nickpell the Payment Record interface seems to use the model `PaymentRecord`: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/accounting/interface.py#L771. Is there a difference? I don't see `BillingRecord` used in the interfaces file

**Comment by nickpell** [comment]
@benrudolph Good point, I misspoke about that.  nvm

**Comment by benrudolph** [comment]
screencapture-localhost-8000-a-aspace-settings-project-billing-statements-download-e207f0c83d4d16657373eec410c6190e-pdf-1430404149646
new pdf (just leaves out the table when doing a wire invoice)

**Comment by benrudolph** [comment]
:+1:

**Comment by nickpell** [comment]
:tada:

**Pull request merged**